### PR TITLE
Upgrade to the latest development incrementalmerkletree & bridgetree versions.

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

**Review by str4d — COMMENTED** [review]
Reviewed everything except `src/rust/src/incremental_merkle_tree.rs`.

File: src/rust/src/wallet.rs [hunk 1]
```diff
@@ -149,11 +150,11 @@ pub struct Wallet {
     nullifiers: BTreeMap<Nullifier, OutPoint>,
     /// The incremental Merkle tree used to track note commitments and witnesses for notes
     /// belonging to the wallet.
-    witness_tree: BridgeTree<MerkleHashOrchard, MERKLE_DEPTH>,
+    commitment_tree: BridgeTree<MerkleHashOrchard, u32, MERKLE_DEPTH>,
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/rust/src/wallet.rs]: I finally remembered the rationale for not using `BlockHeight`: https://github.com/zcash/zcash/pull/6549/files#diff-6dd0819ddbce2e6255b02b0dd4dae639762d8ad95c57879be3ab766b3be08ed7R171-R174

File: .cargo/config.offline [hunk 2]
```diff
@@ -4,9 +4,19 @@ linker = "aarch64-linux-gnu-gcc"
...
+
 [source."https://github.com/zcash/librustzcash.git"]
 git = "https://github.com/zcash/librustzcash.git"
-rev = "ec1c1bcf21c43b0736d298d3a437162253946a8a"
+rev = "d56c6624a61218cdb56a3d93c5f6c30287606120"
+replace-with = "vendored-sources"
+
+[source."https://github.com/zcash/orchard.git"]
+git = "https://github.com/zcash/orchard.git"
+rev = "11b6858ac81a79675de205980155a49d6f92b71e"
```
Comment (daira) [inline .cargo/config.offline:19]: I have not checked these hashes.

File: src/rust/include/rust/orchard/wallet.h [hunk 3]
```diff
@@ -366,7 +368,7 @@ OrchardSpendInfoPtr* orchard_wallet_get_spend_info(
         const OrchardWalletPtr* wallet,
         const unsigned char txid[32],
         uint32_t action_idx,
-        const unsigned char as_of_root[32]);
+        size_t confirmations);
```
Comment (daira) [inline src/rust/include/rust/orchard/wallet.h:371]: Update the comment to say something like "relative to the root of the Orchard commitment tree with the given number of confirmations."

File: src/rust/include/rust/orchard/wallet.h [hunk 4]
```diff
@@ -174,10 +174,12 @@ bool orchard_wallet_append_bundle_commitments(
         );
 
 /**
- * Returns the root of the wallet's note commitment tree.
+ * Returns the root of the wallet's note commitment tree by assigning to `root_ret`.
+ * Returns `true`
```
Comment (daira) [inline src/rust/include/rust/orchard/wallet.h:178]:
```suggestion
Returns the root of the wallet's Orchard note commitment tree with the given number
of confirmations.
```
Also say what conditions could cause a false return.

File: src/rust/src/incremental_merkle_tree.rs [hunk 5]
```diff
@@ -1,120 +1,276 @@
...
     let checkpoints = match ser_version {
         SER_V1 => {
-            let mut save_positions = saved.iter().map(|(p, i)| (*i, *p)).collect();
-            Vector::read_collected_mut(&mut reader, |r| read_checkpoint_v1(r, &mut save_positions))
+            // SER_V1 checkpoint serialization encoded checkpoint data from the `Checkpoint` type
+            // as defined in `incrementalmerkletree` version `0.3.0-beta-2`. This version was only
+            // used in testnet wallets prior to NU5 launch. As such, we no longer support reading
+            // v1 checkpoint data.
+            Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
```
Comment (daira) [inline src/rust/src/incremental_merkle_tree.rs:259]:
```suggestion
                io::ErrorKind::InvalidData,
```
https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidData says:
> Unlike [`InvalidInput`](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidInput), this typically means that the operation parameters were valid, however the error was caused by malformed input data.
which is correct here.

File: src/rust/src/incremental_merkle_tree.rs [hunk 6]
```diff
@@ -1,120 +1,276 @@
...
+            let mut fake_checkpoint_id = 0u32;
+            Vector::read_collected_mut(&mut reader, |r| {
+                fake_checkpoint_id += 1;
+                read_checkpoint_v2(r, fake_checkpoint_id)
+            })
         }
-        SER_V2 => Vector::read_collected(&mut reader, |r| read_checkpoint_v2(r)),
+        SER_V3 => Vector::read_collected(&mut reader, |r| read_checkpoint_v3(r)),
         flag => Err(io::Error::new(
             io::ErrorKind::InvalidInput,
```
Comment (daira) [inline src/rust/src/incremental_merkle_tree.rs:275]:
```suggestion
            io::ErrorKind::InvalidData,
```

File: src/rust/src/merkle_frontier.rs [hunk 7]
```diff
@@ -67,16 +69,16 @@ impl<H: Copy + Hashable + HashSer> MerkleFrontier<H> {
         root
     }
 
-    /// Returns the number of leaves appended to the frontier.
+    /// Returns the number of leaves in the Merkle tree for which this is the frontier.
```
Comment (daira) [inline src/rust/src/merkle_frontier.rs:72]: Is this a change in semantics?

File: src/rust/src/wallet.rs [hunk 8]
```diff
@@ -940,14 +941,20 @@ pub extern "C" fn orchard_wallet_append_bundle_commitments(
...
     wallet: *const Wallet,
+    checkpoint_depth: usize,
     root_ret: *mut [u8; 32],
-) {
+) -> bool {
     let wallet = unsafe { wallet.as_ref() }.expect("Wallet pointer may not be null");
     let root_ret = unsafe { root_ret.as_mut() }.expect("Cannot return to the null pointer.");
 
     // there is always a valid note commitment tree root at depth 0
     // (it may be the empty root)
```
Comment (daira) [inline src/rust/src/wallet.rs:951]:
```suggestion
```
(comment is no longer applicable, even if it is still a true statement)

File: src/rust/include/rust/orchard/wallet.h [hunk 9]
```diff
@@ -174,10 +174,12 @@ bool orchard_wallet_append_bundle_commitments(
...
  */
-void orchard_wallet_commitment_tree_root(
+bool orchard_wallet_commitment_tree_root(
         const OrchardWalletPtr* wallet,
+        const size_t confirmations,
```
Comment (daira) [inline src/rust/include/rust/orchard/wallet.h:182]: Why is this called `confirmations`, when it is called `checkpoint_depth` in the Rust code?

File: src/wallet/gtest/test_orchard_wallet.cpp [hunk 10]
```diff
@@ -110,7 +110,7 @@ TEST(TransactionBuilder, OrchardToOrchard) {
 
     // If we attempt to get spend info now, it will fail because the note hasn't
     // been witnessed in the Orchard commitment tree.
-    EXPECT_THROW(wallet.GetSpendInfo(notes, wallet.GetLatestAnchor()), std::logic_error);
+    EXPECT_THROW(wallet.GetSpendInfo(notes, 1, wallet.GetLatestAnchor()), std::logic_error);
```
Comment (daira) [inline src/wallet/gtest/test_orchard_wallet.cpp:113]: Why is this 1 rather than 0?

File: src/wallet/gtest/test_orchard_wallet.cpp [hunk 11]
```diff
@@ -119,7 +119,7 @@ TEST(TransactionBuilder, OrchardToOrchard) {
     ASSERT_TRUE(wallet.AppendNoteCommitments(2, fakeBlock));
 
     // Now we can get spend info for the note.
-    auto spendInfo = wallet.GetSpendInfo(notes, wallet.GetLatestAnchor());
+    auto spendInfo = wallet.GetSpendInfo(notes, 1, wallet.GetLatestAnchor());
```
Comment (daira) [inline src/wallet/gtest/test_orchard_wallet.cpp:122]: Why is this 1 rather than 0?

File: src/wallet/gtest/test_orchard_wallet.cpp (hunk 10 above)
Comment (daira) [reply src/wallet/gtest/test_orchard_wallet.cpp:113]: Oh I see, `OrchardWallet::GetSpendInfo` uses `anchorConfirmations - 1`.

Hmm, it that seems potentially error prone to use different conventions at the different layers, though.

File: src/rust/include/rust/orchard/wallet.h (hunk 9 above)
Comment (daira) [reply src/rust/include/rust/orchard/wallet.h:182]: Bearing in mind https://github.com/zcash/zcash/pull/6549/commits/be0f0bf95b0be176177afe33268ad5dcb65d8bd9#r1209484007, this looks simply wrong. The checkpoint depth and the number of confirmations are not the same thing; they differ by 1.
```suggestion
        const size_t checkpoint_depth,
```

File: src/wallet/orchard.h [hunk 12]
```diff
@@ -346,10 +346,22 @@ class OrchardWallet
 
     uint256 GetLatestAnchor() const {
         uint256 value;
-        orchard_wallet_commitment_tree_root(inner.get(), value.begin());
+        // there is always a valid note commitment tree root at depth 0
+        assert(orchard_wallet_commitment_tree_root(inner.get(), 0, value.begin()));
         return value;
     }
 
+    std::optional<uint256> GetAnchorWithConfirmations(unsigned int confirmations) const {
```
Comment (daira) [inline src/wallet/orchard.h:354]: Document that `confirmations` must be at least 1.

File: src/wallet/orchard.h [hunk 13]
```diff
@@ -445,7 +457,8 @@ class OrchardWallet
 
     std::vector<std::pair<libzcash::OrchardSpendingKey, orchard::SpendInfo>> GetSpendInfo(
```
Comment (daira) [inline src/wallet/orchard.h:458]: Document that `anchorConfirmations` must be at least 1.

File: src/wallet/wallet.h [hunk 14]
```diff
@@ -2008,7 +2008,8 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
          uint256 &final_anchor) const;
     std::vector<std::pair<libzcash::OrchardSpendingKey, orchard::SpendInfo>> GetOrchardSpendInfo(
```
Comment (daira) [inline src/wallet/wallet.h:2009]: Document that `confirmations` must be at least 1.

File: src/wallet/orchard.cpp [hunk 15]
```diff
@@ -14,15 +14,19 @@ std::optional<libzcash::OrchardSpendingKey> OrchardWallet::GetSpendingKeyForAddr
 
 std::vector<std::pair<libzcash::OrchardSpendingKey, orchard::SpendInfo>> OrchardWallet::GetSpendInfo(
     const std::vector<OrchardNoteMetadata>& noteMetadata,
-    uint256 anchor) const
+    unsigned int anchorConfirmations,
+    const uint256& anchor) const
```
Comment (daira) [inline src/wallet/orchard.cpp:18]: Why do we require `anchor` to be passed in when, for a given `anchorConfirmations`, there is only one value it could have?

**Review by daira — COMMENTED** [review]
Reviewed with suggestions and some questions.

File: src/rust/src/merkle_frontier.rs (hunk 7 above)
Comment (nuttycom) [reply src/rust/src/merkle_frontier.rs:72]: No, this has the same meaning, it's just clarifying (since the previous comment there didn't specify "appended to since when.")

File: src/wallet/orchard.cpp (hunk 15 above)
Comment (nuttycom) [reply src/wallet/orchard.cpp:18]: I continue to pass in `anchor` here as a sanity check; we obtain `anchor` from the coins db, which caches the Merkle frontier for each chain state, and here we verify that against the Merkle root computed from the wallet's `BridgeTree` instance. Previously this check was implicit, because we used the anchor to search for the tree state that we're interested in. Performing that search required recomputing each candidate root of the `BridgeTree`; now we no longer have to do that, but I think it's useful to still compute it once to make certain that we're using the correct Merkle path when making the proofs.

File: src/rust/src/wallet.rs (hunk 8 above)
Comment (nuttycom) [reply src/rust/src/wallet.rs:951]: I think that this comment is still relevant; `checkpoint_depth` will be 0, it's `anchorConfirmations` that must be >= 1.

**Comment by nuttycom** [comment]
[force-pushed](https://github.com/zcash/zcash/compare/cae2c5f467c7e28bff4eea52b1f3fd75de157139..64ae2d9) to address comments from code review.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -3925,10 +3926,11 @@ bool CWallet::GetSaplingNoteWitnesses(const std::vector<SaplingOutPoint>& notes,
 
 std::vector<std::pair<libzcash::OrchardSpendingKey, orchard::SpendInfo>> CWallet::GetOrchardSpendInfo(
     const std::vector<OrchardNoteMetadata>& orchardNoteMetadata,
-    uint256 anchor) const
+    unsigned int confirmations,
```
Comment (daira) [inline src/wallet/wallet.cpp:3929]:
```suggestion
    unsigned int anchorConfirmations,
```

File: src/wallet/wallet.cpp [hunk 17]
```diff
@@ -3925,10 +3926,11 @@ bool CWallet::GetSaplingNoteWitnesses(const std::vector<SaplingOutPoint>& notes,
...
+    const uint256& anchor) const
 {
     AssertLockHeld(cs_wallet);
-    return orchardWallet.GetSpendInfo(orchardNoteMetadata, anchor);
+    return orchardWallet.GetSpendInfo(orchardNoteMetadata, confirmations, anchor);
```
Comment (daira) [inline src/wallet/wallet.cpp:3933]:
```suggestion
    return orchardWallet.GetSpendInfo(orchardNoteMetadata, anchorConfirmations, anchor);
```

**Review by daira — APPROVED** [review]
utACK

**Review by ebfull — APPROVED** [review]

**Review by ebfull — DISMISSED** [review]
Suggested some improvements in https://github.com/nuttycom/zcash/pull/4 but I'm ACK on this PR as-is.

**Review by ebfull — DISMISSED** [review]

**Review by daira — DISMISSED** [review]
utACK

**Comment by nuttycom** [comment]
[force-pushed](https://github.com/zcash/zcash/compare/1fdb67c361ceebf5df44743ee8094089926b8eb0..85148bc315ae03422993d9229ad0535b0f2142a2) to rebase on fixes for nondeterministic test failures on `master`.

**Review by ebfull — APPROVED** [review]

File: qa/supply-chain/imports.lock [hunk 18]
```diff
@@ -201,12 +217,52 @@ who = "Johan Andersson <opensource@embark-studios.com>"
 criteria = "safe-to-deploy"
 version = "1.0.58"
 
+[[audits.embark-studios.audits.epaint]]
+who = "Johan Andersson <opensource@embark-studios.com>"
+criteria = "safe-to-deploy"
+violation = "<0.20.0"
+notes = "Specified crate license does not include licenses of embedded fonts if using default features or the `default_fonts` feature. Tracked in: https://github.com/emilk/egui/issues/2321"
```
Comment (daira) [inline qa/supply-chain/imports.lock:224]: I do not see `epaint` in `Cargo.lock`, so why is this here? If we did depend on `epaint`, we'd need to include the OFL-1.0 and UFL-1.0 licenses for these fonts in `contrib/debian/copyright`.
